src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx

```diff
@@ -1267,12 +1282,40 @@ cpdef object compute_network_structured(
                   flowveldepth[segment.id, timestep, 1] = out_buf[i, 1]
                   flowveldepth[segment.id, timestep, 2] = out_buf[i, 2]
                     
-        if gages_size:
+        if gages_size:  # TODO: This loops over all gages for all reaches.
+                        # We should have a membership test at the reach loop level
+                        # so that we only enter this process for reaches where the
+                        # gage actually exists. We have the filter in place to
+                        # filter the gage list so that only relevant gages for a
+                        # particular network are present in the function call ---
+                        # adding the reach-based filter would be the next level.
+
             for gage_i in range(gages_size):
-                usgs_position_i = usgs_positions_list[gage_i]
-                if timestep < gage_maxtimestep:
+                usgs_position_i = usgs_positions[gage_i] 
+                # TODO: It is possible to remove the following branching logic if
+                # we just loop over the timesteps during DA and post-DA, if that
+                # is a major performance optimization. On the flip side, it would
+                # probably introduce unwanted code complexity.
+                if (timestep < gage_maxtimestep and not isnan(usgs_values[gage_i,timestep-1])):
                     flowveldepth[usgs_position_i, timestep, 0] = usgs_values[gage_i, timestep-1]
-               
+                    # add/update lastobs_timestep
+                    lastobs_timestep[gage_i] = timestep - 1
+                    lastobs_values[gage_i] = usgs_values[gage_i, timestep-1]
+                else:
+                    a = 120  # TODO: pull this a value from the config file somehow
+                    if lastobs_timestep[gage_i] < 0: # Initialized to -1
+                        da_decay_minutes = (timestep) * dt / 60 - time_since_lastobs_init[gage_i] # seconds to minutes
+                    else:
+                        da_decay_minutes = (timestep - lastobs_timestep[gage_i]) * dt / 60
+                    da_weight = exp(da_decay_minutes/-a)  # TODO: This could be pre-calculated knowing when obs finish relative to simulation time
+                    # replacement_value = f(lastobs_value, da_weight)  # TODO: we need to be able to export these values to compute the 'Nudge'
+                    da_shift = lastobs_values[gage_i] - flowveldepth[usgs_position_i, timestep, 0]
+                    da_weighted_shift = da_shift * da_weight
+                    replacement_value = flowveldepth[usgs_position_i, timestep, 0] + da_weighted_shift
+                    # printf("decaying from timestep: %d %d\t", timestep, gages_size)
+                    flowveldepth[usgs_position_i, timestep, 0] = replacement_value
```

Comment:
This block should be a function, `obs_persist_da` which could then be tested with the test sequences in the PR explanation.

Reply:
The function has been added.